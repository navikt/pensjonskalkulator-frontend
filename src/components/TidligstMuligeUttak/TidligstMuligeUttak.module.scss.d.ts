import globalClassNames from '../../style.d'
declare const classNames: typeof globalClassNames & {
  readonly loader: 'loader'
  readonly paragraph: 'paragraph'
}
export = classNames
