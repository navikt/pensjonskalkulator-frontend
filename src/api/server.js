// const express = require('express')
// const liveness = require('../src/__mocks__/data/liveness.json')
// const pensjonsberegning = require('../src/__mocks__/data/pensjonsberegning.json')

// const app = express()
// const port = 8088

// app.get('/api/status', (request, response) => {
//   response.send(liveness)
// })

// app.get('/api/pensjonsberegning', async (request, response) => {
//   response.send(pensjonsberegning)
// })

// app.listen(process.env.PORT || port, () => {
//   console.info('proxy server is running on port', process.env.PORT || port)
// })

// import { rest } from 'msw'
// import { setupServer } from 'msw/node'

import { setupServer } from 'msw/node'
import { handlers } from './handlers.js'

export const server = setupServer(...handlers)
server.listen({ onUnhandledRequest: 'error' })
