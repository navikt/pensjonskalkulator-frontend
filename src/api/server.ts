import { setupServer } from 'msw/node'
import { getHandlers } from './handlers.js'
import { rest } from 'msw'

export const API_TARGET = 'http://localhost:8088'
export const API_PATH = '/pensjon/kalkulator/api'

const handlers = getHandlers(`${API_TARGET}${API_PATH}`)
export const server = setupServer(...handlers)

server.listen({ onUnhandledRequest: 'error' })

type MockResponseOptions = {
  status?: number
  json?: any
}

export const mockResponse = (
  path: string,
  options: MockResponseOptions = {
    status: 200,
    json: 'OK',
  }
) => {
  server.use(
    rest.get(`${API_TARGET}${API_PATH}${path}`, (req, res, ctx) => {
      return res(ctx.status(options.status ?? 200), ctx.json(options.json))
    })
  )
}

export const mockErrorResponse = (
  path: string,
  options: MockResponseOptions = {
    status: 500,
    json: "Beep boop I'm an error!",
  }
) => {
  mockResponse(path, {
    ...options,
    status: options.status ?? 500,
  })
}
