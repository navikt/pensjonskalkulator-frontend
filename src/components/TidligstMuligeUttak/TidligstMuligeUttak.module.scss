@import '../../scss/variables';

.section {
  composes: whitesection from '../../scss/modules/whitesection.module.scss';
  display: flex;
  flex-direction: column;
  gap: var(--a-spacing-4);
  padding: var(--a-spacing-6);
  text-align: start;
  margin-bottom: var(--a-spacing-6);

  @media (min-width: $a-breakpoint-sm) {
    padding: var(--a-spacing-10) var(--a-spacing-10);
  }
}

.loader {
  margin: var(--a-spacing-6) auto;
}

.chipsWrapper {
  /* double-specificity n√∏dvendig pga Aksel overskrivning */
  &.chipsWrapper__hasGap {
    gap: var(--a-spacing-4);
  }

  :global {
    .navds-chips--icon-left {
      padding-left: var(--a-spacing-3);
    }

    .navds-chips__toggle-icon {
      display: none;
    }
  }
}

.readMore__hasPadding {
  margin-bottom: var(--a-spacing-3);
}

.readMore__isCentered {
  margin: 0 auto;
}

.paragraph {
  max-width: 75ch;
}
