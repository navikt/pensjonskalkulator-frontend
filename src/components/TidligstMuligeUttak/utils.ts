export const generateAlderArray = (
  startAlder: number,
  endAlder: number,
  formatertTidligsteMuligeUttaksalder: string
) => {
  const alderArray: string[] = []
  for (let i = startAlder; i <= endAlder; i++) {
    if (i === startAlder) {
      alderArray.push(formatertTidligsteMuligeUttaksalder)
    } else {
      alderArray.push(`${i.toString()} 책r`)
    }
  }
  return alderArray
}

export const generateXAxis = (startAlder: number, endAlder: number) => {
  const alderArray: string[] = []
  for (let i = startAlder; i <= endAlder; i++) {
    if (i === startAlder) {
      alderArray.push((i - 1).toString())
    }
    alderArray.push(i.toString())
  }
  return alderArray
}

export const formatUttaksalder = (
  { aar, maaned }: Uttaksalder,
  options: { compact: boolean } = { compact: false }
): string => {
  return maaned !== 0
    ? `${aar} 책r og ${maaned} ${options.compact ? 'md.' : 'm책neder'}`
    : `${aar} 책r`
}
