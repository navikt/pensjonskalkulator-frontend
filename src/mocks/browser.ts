import { setupWorker } from 'msw/browser'

import { getHandlers } from './handlers'

const handlers = getHandlers()
export const worker = setupWorker(...handlers)
