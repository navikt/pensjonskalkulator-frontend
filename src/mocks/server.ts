import { rest } from 'msw'
import { setupServer } from 'msw/node'

import { API_BASEURL } from '@/paths'

import { getHandlers } from './handlers'

const handlers = getHandlers(API_BASEURL)
export const server = setupServer(...handlers)

type MockResponseOptions = {
  status?: number
  json?: ReturnType<typeof JSON.parse>
  text?: string
  method?: 'post' | 'get'
  baseUrl?: string
}

const defaultResponseOptions: Required<MockResponseOptions> = {
  status: 200,
  json: { data: 'OK' },
  method: 'get',
  baseUrl: `${API_BASEURL}`,
  text: '',
}

export const mockResponse = (
  path: string,
  inputOptions: MockResponseOptions = {}
) => {
  const options = { ...defaultResponseOptions, ...inputOptions }
  const parsedPath = `${options.baseUrl}${path}`

  server.use(
    rest[options.method](parsedPath, (_req, res, ctx) => {
      return res(
        ctx.status(options.status),
        options.text ? ctx.text(options.text) : ctx.json(options.json)
      )
    })
  )
}

const defaultErrorResponseOptions: Required<MockResponseOptions> = {
  status: 500,
  json: { data: "Beep boop I'm an error!" },
  text: '',
  method: 'get',
  baseUrl: `${API_BASEURL}`,
}

export const mockErrorResponse = (
  path: string,
  inputOptions: MockResponseOptions = {}
) => {
  const options = {
    ...defaultErrorResponseOptions,
    ...inputOptions,
  }
  mockResponse(path, options)
}
