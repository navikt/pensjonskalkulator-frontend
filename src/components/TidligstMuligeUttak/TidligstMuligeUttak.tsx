import React, { useMemo, useState } from 'react'

import {
  Alert,
  BodyLong,
  Chips,
  Heading,
  Loader,
  ReadMore,
} from '@navikt/ds-react'
import clsx from 'clsx'

import { Pensjonssimulering } from '../Pensjonssimulering'
import {
  formatUttaksalder,
  generateAlderArray,
} from '../TidligstMuligeUttak/utils'
import { Grunnlag } from '@/components/Grunnlag'
import { useGetTidligsteMuligeUttaksalderQuery } from '@/state/api/apiSlice'

import styles from './TidligstMuligeUttak.module.scss'

const useAlderChips = (data?: Uttaksalder, maksalder = 77): string[] =>
  useMemo(
    () =>
      data?.aar
        ? generateAlderArray(
            data.aar,
            maksalder,
            formatUttaksalder(data, { compact: true })
          )
        : [],
    [data]
  )

export function TidligstMuligeUttak() {
  const {
    data: tidligstMuligUttak,
    isLoading,
    isError,
    isSuccess,
  } = useGetTidligsteMuligeUttaksalderQuery()
  const [valgtUttaksalder, setValgtUttaksalder] = useState<string | undefined>()

  const alderChips = useAlderChips(tidligstMuligUttak)

  if (isLoading) {
    return (
      <Loader
        className={styles.loader}
        data-testid="loader"
        size="3xlarge"
        title="Henter tidligste mulige uttaksalder"
      />
    )
  }

  if (isError || !isSuccess) {
    return (
      <Alert variant="error">
        <Heading spacing size="small" level="2">
          Vi klarte ikke å hente din tidligste mulige uttaksalder. Prøv igjen
          senere.
        </Heading>
      </Alert>
    )
  }

  return (
    <>
      <section className={styles.section}>
        <BodyLong className={styles.paragraph}>
          Du kan tidligst ta ut alderspensjon når du er{' '}
          {formatUttaksalder(tidligstMuligUttak)}. Hvis du går av senere, får du
          høyere pensjon i året.
        </BodyLong>
        <ReadMore header="Hva avgjør tidligste uttakstidspunkt?">
          {'//TODO'}
        </ReadMore>
      </section>

      <section className={styles.section}>
        <Heading size="xsmall" level="2">
          Når vil du ta ut alderspensjon?
        </Heading>
        <Chips
          className={clsx(styles.chipsWrapper, styles.chipsWrapper__hasGap)}
        >
          {alderChips.length > 0 &&
            alderChips.slice(0, 6).map((alderChip) => (
              <Chips.Toggle
                selected={valgtUttaksalder === alderChip}
                key={alderChip}
                onClick={() => setValgtUttaksalder(alderChip)}
              >
                {alderChip}
              </Chips.Toggle>
            ))}
        </Chips>
        <ReadMore
          header="Vis flere aldere"
          className={clsx({ [styles.readMore__hasPadding]: valgtUttaksalder })}
        >
          <Chips
            className={clsx(styles.chipsWrapper, styles.chipsWrapper__hasGap)}
          >
            {alderChips.length > 0 &&
              alderChips.slice(6, alderChips.length).map((alderChip) => (
                <Chips.Toggle
                  selected={valgtUttaksalder === alderChip}
                  key={alderChip}
                  onClick={() => setValgtUttaksalder(alderChip)}
                >
                  {alderChip}
                </Chips.Toggle>
              ))}
          </Chips>
        </ReadMore>
        {valgtUttaksalder && (
          <>
            <Pensjonssimulering uttaksalder={parseInt(valgtUttaksalder, 10)} />
            <ReadMore
              header="Vis tabell"
              className={styles.readMore__isCentered}
            >
              {'// TODO'}
            </ReadMore>
          </>
        )}
      </section>

      {valgtUttaksalder && (
        <section className={styles.section}>
          <Grunnlag tidligstMuligUttak={tidligstMuligUttak} />
        </section>
      )}
    </>
  )
}
