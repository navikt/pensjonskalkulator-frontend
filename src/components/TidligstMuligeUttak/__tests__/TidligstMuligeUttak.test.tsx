import { describe, expect, it, vi } from 'vitest'

import { TidligstMuligeUttak } from '../TidligstMuligeUttak'
import { mockErrorResponse, mockResponse } from '@/mocks/server'
import {
  fireEvent,
  render,
  screen,
  swallowErrorsAsync,
  waitFor,
} from '@/test-utils'

describe('TidligstMuligeUttak', () => {
  it('viser loading og deretter riktig header, tekst og knapper', async () => {
    const result = render(<TidligstMuligeUttak />)
    expect(screen.getByTestId('loader')).toBeVisible()

    await waitFor(() => {
      expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent(
        'Når vil du ta ut alderspensjon?'
      )
      expect(
        screen.getByText('Du kan tidligst ta ut alderspensjon', {
          exact: false,
        })
      ).toBeVisible()
      expect(
        screen.getByText('Hva avgjør tidligste uttakstidspunkt?')
      ).toBeVisible()
      const buttons = screen.getAllByRole('button')
      expect(buttons).toHaveLength(8)
      expect(
        screen.queryByRole('button', { pressed: true })
      ).not.toBeInTheDocument()
      fireEvent.click(screen.getByText('Vis flere aldere'))
      const flereButtons = screen.getAllByRole('button')
      expect(flereButtons).toHaveLength(18)
      expect(screen.queryByTestId('loader')).not.toBeInTheDocument()
      expect(result.asFragment()).toMatchSnapshot()
    })
  })

  it('viser feilmelding om henting av pensjonberegning feiler', async () => {
    mockErrorResponse('/tidligste-uttaksalder')

    const result = render(<TidligstMuligeUttak />)

    await waitFor(() => {
      expect(
        screen.getByText(
          'Vi klarte ikke å hente din tidligste mulige uttaksalder. Prøv igjen senere.'
        )
      ).toBeVisible()
      expect(result.asFragment()).toMatchSnapshot()
    })
  })

  it('viser feilmelding om pensjonsberegning er på ugyldig format', async () => {
    const invalidData = {
      aar: 67,
      maaned: null,
    } as unknown as Uttaksalder
    mockResponse('/tidligste-uttaksalder', { json: [invalidData] })

    render(<TidligstMuligeUttak />)

    await swallowErrorsAsync(async () => {
      await waitFor(() => {
        expect(
          screen.getByText(
            'Vi klarte ikke å hente din tidligste mulige uttaksalder. Prøv igjen senere.'
          )
        ).toBeVisible()
      })
    })
  })

  it('oppdaterer valgt knapp og tegner graph når brukeren klikker på en knapp', async () => {
    const { container } = render(<TidligstMuligeUttak />)

    await waitFor(async () => {
      await fireEvent.click(screen.getByText('Vis flere aldere'))

      await fireEvent.click(screen.getByText('72 år'))
      expect(screen.getByRole('button', { pressed: true })).toHaveTextContent(
        '72 år'
      )
      vi.useFakeTimers()
      vi.advanceTimersByTime(250)
      expect(
        container.getElementsByClassName('highcharts-container').length
      ).toBe(1)
      expect(screen.getByText('Vis tabell')).toBeVisible()
      vi.useRealTimers()
    })
  })
})
