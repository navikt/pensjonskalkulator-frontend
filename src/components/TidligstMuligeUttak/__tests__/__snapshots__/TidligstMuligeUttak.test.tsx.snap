// Vitest Snapshot v1

exports[`TidligstMuligeUttak > viser feilmelding om henting av pensjonberegning feiler 1`] = `
<DocumentFragment>
  <div
    class="navds-alert navds-alert--error navds-alert--medium"
  >
    <svg
      aria-labelledby="title-r6"
      class="navds-alert__icon"
      fill="none"
      focusable="false"
      height="1em"
      role="img"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title
        id="title-r6"
      >
        Feil
      </title>
      <path
        clip-rule="evenodd"
        d="M12 0c6.627 0 12 5.373 12 12s-5.373 12-12 12S0 18.627 0 12 5.373 0 12 0Zm3.571 7L17 8.429 13.428 12 17 15.571 15.571 17 12 13.428 8.429 17 7 15.571 10.572 12 7 8.429 8.429 7 12 10.572 15.571 7Z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
    <div
      class="navds-alert__wrapper navds-body-long"
    >
      <h2
        class="navds-heading navds-heading--small navds-typo--spacing"
      >
        Vi klarte ikke å hente din tidligste mulige uttaksalder. Prøv igjen senere.
      </h2>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TidligstMuligeUttak > viser loading og deretter riktig header, tekst og knapper 1`] = `
<DocumentFragment>
  <section
    class="_section_d3b588"
  >
    <p
      class="_paragraph_d3b588 navds-body-long"
    >
      Du kan tidligst ta ut alderspensjon når du er 62 år og 10 måneder. Hvis du går av senere, får du høyere pensjon i året.
    </p>
    <div
      class="navds-read-more navds-read-more--medium"
    >
      <button
        aria-expanded="false"
        class="navds-read-more__button navds-body-short"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="navds-read-more__expand-icon"
          fill="none"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12 17.414 4.293 9.707l1.414-1.414L12 14.586l6.293-6.293 1.414 1.414L12 17.414Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
        <svg
          aria-hidden="true"
          class="navds-read-more__expand-icon navds-read-more__expand-icon--filled"
          fill="none"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="m12 18.121-8.06-8.06 2.12-2.122L12 13.88l5.94-5.94 2.12 2.122L12 18.12Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
        <span>
          Hva avgjør tidligste uttakstidspunkt?
        </span>
      </button>
      <div
        style="height: 0px; overflow: hidden;"
      >
        <div
          aria-hidden="true"
          style="display: none;"
        >
          <div
            class="navds-read-more__content navds-body-long"
          >
            //TODO
          </div>
        </div>
      </div>
    </div>
  </section>
  <section
    class="_section_d3b588"
  >
    <h2
      class="navds-heading navds-heading--xsmall"
    >
      Når vil du ta ut alderspensjon?
    </h2>
    <ul
      class="navds-chips _chipsWrapper_d3b588 _chipsWrapper__hasGap_d3b588 navds-chips--medium navds-body-short navds-body-short--small"
    >
      <li>
        <button
          aria-pressed="false"
          class="navds-chips__chip navds-chips__toggle"
        >
          <span
            class="navds-chips__chip-text"
          >
            62 år og 10 md.
          </span>
        </button>
      </li>
      <li>
        <button
          aria-pressed="false"
          class="navds-chips__chip navds-chips__toggle"
        >
          <span
            class="navds-chips__chip-text"
          >
            63 år
          </span>
        </button>
      </li>
      <li>
        <button
          aria-pressed="false"
          class="navds-chips__chip navds-chips__toggle"
        >
          <span
            class="navds-chips__chip-text"
          >
            64 år
          </span>
        </button>
      </li>
      <li>
        <button
          aria-pressed="false"
          class="navds-chips__chip navds-chips__toggle"
        >
          <span
            class="navds-chips__chip-text"
          >
            65 år
          </span>
        </button>
      </li>
      <li>
        <button
          aria-pressed="false"
          class="navds-chips__chip navds-chips__toggle"
        >
          <span
            class="navds-chips__chip-text"
          >
            66 år
          </span>
        </button>
      </li>
      <li>
        <button
          aria-pressed="false"
          class="navds-chips__chip navds-chips__toggle"
        >
          <span
            class="navds-chips__chip-text"
          >
            67 år
          </span>
        </button>
      </li>
    </ul>
    <div
      class="navds-read-more navds-read-more--medium navds-read-more--open"
    >
      <button
        aria-expanded="true"
        class="navds-read-more__button navds-body-short"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="navds-read-more__expand-icon"
          fill="none"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12 17.414 4.293 9.707l1.414-1.414L12 14.586l6.293-6.293 1.414 1.414L12 17.414Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
        <svg
          aria-hidden="true"
          class="navds-read-more__expand-icon navds-read-more__expand-icon--filled"
          fill="none"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="m12 18.121-8.06-8.06 2.12-2.122L12 13.88l5.94-5.94 2.12 2.122L12 18.12Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
        <span>
          Vis flere aldere
        </span>
      </button>
      <div
        style="height: 0px; overflow: hidden; transition: height 250ms ease 0ms;"
      >
        <div
          aria-hidden="false"
          style=""
        >
          <div
            class="navds-read-more__content navds-body-long"
          >
            <ul
              class="navds-chips _chipsWrapper_d3b588 _chipsWrapper__hasGap_d3b588 navds-chips--medium navds-body-short navds-body-short--small"
            >
              <li>
                <button
                  aria-pressed="false"
                  class="navds-chips__chip navds-chips__toggle"
                >
                  <span
                    class="navds-chips__chip-text"
                  >
                    68 år
                  </span>
                </button>
              </li>
              <li>
                <button
                  aria-pressed="false"
                  class="navds-chips__chip navds-chips__toggle"
                >
                  <span
                    class="navds-chips__chip-text"
                  >
                    69 år
                  </span>
                </button>
              </li>
              <li>
                <button
                  aria-pressed="false"
                  class="navds-chips__chip navds-chips__toggle"
                >
                  <span
                    class="navds-chips__chip-text"
                  >
                    70 år
                  </span>
                </button>
              </li>
              <li>
                <button
                  aria-pressed="false"
                  class="navds-chips__chip navds-chips__toggle"
                >
                  <span
                    class="navds-chips__chip-text"
                  >
                    71 år
                  </span>
                </button>
              </li>
              <li>
                <button
                  aria-pressed="false"
                  class="navds-chips__chip navds-chips__toggle"
                >
                  <span
                    class="navds-chips__chip-text"
                  >
                    72 år
                  </span>
                </button>
              </li>
              <li>
                <button
                  aria-pressed="false"
                  class="navds-chips__chip navds-chips__toggle"
                >
                  <span
                    class="navds-chips__chip-text"
                  >
                    73 år
                  </span>
                </button>
              </li>
              <li>
                <button
                  aria-pressed="false"
                  class="navds-chips__chip navds-chips__toggle"
                >
                  <span
                    class="navds-chips__chip-text"
                  >
                    74 år
                  </span>
                </button>
              </li>
              <li>
                <button
                  aria-pressed="false"
                  class="navds-chips__chip navds-chips__toggle"
                >
                  <span
                    class="navds-chips__chip-text"
                  >
                    75 år
                  </span>
                </button>
              </li>
              <li>
                <button
                  aria-pressed="false"
                  class="navds-chips__chip navds-chips__toggle"
                >
                  <span
                    class="navds-chips__chip-text"
                  >
                    76 år
                  </span>
                </button>
              </li>
              <li>
                <button
                  aria-pressed="false"
                  class="navds-chips__chip navds-chips__toggle"
                >
                  <span
                    class="navds-chips__chip-text"
                  >
                    77 år
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
